source/fe/fe_enriched.cc

```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+    void
+    make_colorwise_enrichment_functions(
+      const unsigned int &                                    num_colors,
+      const std::vector<std::shared_ptr<Function<spacedim>>> &enrichments,
+      const std::map<unsigned int, std::map<unsigned int, unsigned int>>
+        &cellwise_color_predicate_map,
+      std::vector<std::function<const Function<spacedim> *(
+        const typename Triangulation<dim, spacedim>::cell_iterator &)>>
+        &color_enrichments)
+    {
+      color_enrichments.clear();
+
+      // Each color should be associated with a single enrichment function
+      // called color enrichment function which calls the correct enrichment
+      // function for a given cell.
+      //
+      // Assume that a cell has a active predicates with ids 4 (color = 1) and
+      // 5 (color = 2). cellwise_color_predicate_map has this information
+      // provided we know the material id.
+      //
+      // Now a call to color_enrichment[1](cell) should in turn call
+      // enrichments[4](cell). That is just what this lambda is doing.
+      color_enrichments.resize(num_colors);
+      for (unsigned int i = 0; i < num_colors; ++i)
+        {
+          color_enrichments[i] =
+            [&, i](const typename Triangulation<dim, spacedim>::cell_iterator
+                     &cell) {
+              unsigned int id = cell->material_id();
+
+              /*
+               * i'th color_enrichment function corresponds to (i+1)'th color.
+               * Since FE_Enriched takes function pointers, we return a
+               * function pointer.
+               */
+              return enrichments[cellwise_color_predicate_map.at(id).at(i + 1)]
+                .get();
+            };
+        }
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    make_fe_collection_from_colored_enrichments(
+      const unsigned int &                       num_colors,
+      const std::vector<std::set<unsigned int>> &fe_sets,
+      const std::vector<std::function<const Function<spacedim> *(
+        const typename Triangulation<dim, spacedim>::cell_iterator &)>>
+        &                              color_enrichments,
+      const FE_Q<dim, spacedim> &      fe_base,
+      const FE_Q<dim, spacedim> &      fe_enriched,
+      const FE_Nothing<dim, spacedim> &fe_nothing,
+      hp::FECollection<dim, spacedim> &fe_collection)
+    {
+      // define dummy function which is associated with FE_Nothing
+      using cell_function = std::function<const Function<spacedim> *(
+        const typename Triangulation<dim, spacedim>::cell_iterator &)>;
+      cell_function dummy_function;
+      dummy_function =
+        [=](const typename Triangulation<dim, spacedim>::cell_iterator &)
+        -> const Function<spacedim> * {
+        AssertThrow(false,
+                    ExcMessage("Called enrichment function for FE_Nothing"));
+        return nullptr;
+      };
+
+      using EnrichmentFunctions_2DVector =
```

Comment:
there is only one place where you use it, so i would not add that `using` shortcut.